File: 153_Find_Minimum_in_Rotated_Sorted_Array/Solution.py
class Solution:
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        if n == 0:
            return 0
        
        if n == 1:
            return nums[0]
        
        if nums[0]<nums[-1]:
            return nums[0]
        
        
        l = 0
        r = n-1
        mid = 0
        while l<=r:
            mid = l+(r-l)//2
            if mid == 0:
                if nums[mid] > nums[mid+1]:
                    return mid+1
                else:
                    l = mid + 1
            elif mid == n-1:
                if nums[mid]<nums[mid-1]:
                    return nums[mid]
                else:
                    r = mid - 1
            elif nums[mid]>nums[mid-1] and nums[mid]>nums[mid+1]:
                return nums[mid+1]
            elif nums[mid] > nums[0]:
                l = mid + 1
            elif nums[mid] < nums[0]:
                r = mid -1
        return nums[mid]
s = Solution()
# s.findMin([3,4,5,1,2])
s.findMin([9,0,2,7,8])

File: 153_Find_Minimum_in_Rotated_Sorted_Array/Solution.cc
#include <iostream>
#include <sstream>
#include <vector>

using namespace std;

class Solution {
public:
    static int findMin(vector<int>& nums) {
        int n = nums.size();
        if(n == 0) return -5001;
        if(n == 1) return nums[0];

        int l=0, r=n-1;
        if(nums[r] > nums[0]) return nums[0];

        while(l<=r){
            int mid = l + (r-l)/2;
            if(mid+1<n && nums[mid]>nums[mid+1]) return nums[mid+1];
            if(mid-1>=0 && nums[mid-1]>nums[mid]) return nums[mid];
            if(nums[mid] > nums[0]) l = mid+1;
            else r = mid -1;
        }
        return -5001;
    }
};

int main(){
    string line;
    while(getline(cin, line)){
        vector<int> nums;
        istringstream is(line);
        int n;
        while(is>>n){
            nums.push_back(n);
        }
        int ret = Solution::findMin(nums);
        cout<<line<<endl;
        cout<<ret<<endl<<endl;
    }
    return 0;
}
